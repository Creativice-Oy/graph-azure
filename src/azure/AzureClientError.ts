import { Response } from "node-fetch";

export default class AzureClientError extends Error {
  private static generateMessage(response: Response): string {
    const intro = "Unexpected response from Azure API:";
    const errorText = `'${response.status} - ${response.statusText}`;

    return [intro, errorText].join(" ");
  }

  public status: number;
  public statusText: string;

  constructor(response: Response) {
    super(AzureClientError.generateMessage(response));
    this.status = response.status;
    this.statusText = response.statusText;
  }
}
