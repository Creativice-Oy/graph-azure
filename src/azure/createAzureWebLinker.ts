import { AzureWebLinker } from "./types";

export default function createAzureWebLinker(
  defaultDomain: string | undefined,
): AzureWebLinker {
  return {
    portalResourceUrl: path => {
      if (defaultDomain && path) {
        return `https://portal.azure.com/#@${defaultDomain}/resource${path}`;
      }
    },
  };
}
