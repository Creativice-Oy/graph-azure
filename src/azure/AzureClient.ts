import querystring from "querystring";

import { IntegrationLogger } from "@jupiterone/jupiter-managed-integration-sdk";

import AzureClientError from "./AzureClientError";
import authenticate from "./graph/authenticate";
import { Group, GroupMember, User } from "./types";

export enum Method {
  GET = "get",
  POST = "post",
}

export interface PaginationOptions {
  limit?: number;
  nextLink?: string;

  /**
   * The property names for `$select` query param.
   */
  select?: string | string[];
}

interface ResourceUrlOptions extends PaginationOptions {
  path: string;
}

interface FetchResourcesInput {
  url: string;
}

export interface FetchResourcesResponse<T extends microsoftgraph.Entity> {
  nextLink: string | undefined;
  resources: T[];
  err?: AzureClientError;
}

export default class AzureClient {
  private clientId: string;
  private clientSecret: string;
  private directoryId: string;
  private accessToken: string;
  private apiHost = "https://graph.microsoft.com";
  private apiVersion = "v1.0";
  private logger: IntegrationLogger;

  constructor(
    clientId: string,
    clientSecret: string,
    directoryId: string,
    logger: IntegrationLogger,
  ) {
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.directoryId = directoryId;
    this.logger = logger;
  }

  public async authenticate() {
    this.accessToken = await authenticate({
      clientId: this.clientId,
      clientSecret: this.clientSecret,
      directoryId: this.directoryId,
    });
  }

  public async fetchUsers(
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<User>> {
    return this.fetchResources({
      url: this.resourceUrl({ path: "/users", ...options }),
    });
  }

  public async fetchGroups(
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<Group>> {
    return this.fetchResources({
      url: this.resourceUrl({ path: "/groups", ...options }),
    });
  }

  public async fetchGroupMembers(
    groupId: string,
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<GroupMember>> {
    return this.fetchResources({
      url: this.resourceUrl({
        path: `/groups/${groupId}/members`,
        ...options,
      }),
    });
  }

  private async fetchResources<T>(
    input: FetchResourcesInput,
  ): Promise<FetchResourcesResponse<T>> {
    try {
      this.logger.trace({ url: input.url }, "Fetching Azure resources");

      const response = await this.makeRequest<any>(input.url, Method.GET);

      return {
        nextLink: response["@odata.nextLink"],
        resources: response.value,
      };
    } catch (err) {
      if (err.status === 404) {
        return { nextLink: undefined, resources: [] };
      } else {
        this.logger.warn(
          { err, url: input.url, method: Method.GET },
          "Azure resource request failed",
        );

        return { nextLink: undefined, resources: [], err };
      }
    }
  }

  /**
   * Pagination: https://docs.microsoft.com/en-us/graph/paging
   * Throttling with retry after: https://docs.microsoft.com/en-us/graph/throttling
   * Batching requests: https://docs.microsoft.com/en-us/graph/json-batching
   */
  private async makeRequest<T>(
    url: string,
    method: Method,
    headers?: {},
  ): Promise<T> {
    if (!this.accessToken) {
      await this.authenticate();
    }

    const options: RequestInit = {
      method,
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
        Authorization: `Bearer ${this.accessToken}`,
        ...headers,
      },
    };

    const response = await fetch(url, options);
    if (response.status >= 300) {
      throw new AzureClientError(response);
    }

    return response.json();
  }

  /**
   * Constructs a resource url to include optional query params.
   *
   * `fetch` will itself perform url escaping, so this function must not do so
   * itself. For example, the `url` library has a nice API for constructing
   * parameters, but the `url.toString()` function answers excaped strings,
   * which `fetch` then attempts to escape, making the params invalid.
   */
  private resourceUrl(options: ResourceUrlOptions): string {
    if (options.nextLink) {
      return options.nextLink;
    } else {
      const queryParams: { [key: string]: any } = {};

      if (options.limit) {
        queryParams.$top = options.limit;
      }

      if (options.select) {
        queryParams.$select = Array.isArray(options.select)
          ? options.select.join(",")
          : options.select;
      }

      let url = `${this.apiHost}/${this.apiVersion}${options.path}`;
      if (Object.keys(queryParams).length > 0) {
        url +=
          "?" +
          querystring.stringify(queryParams, undefined, undefined, {
            encodeURIComponent: (s: string) => s,
          });
      }

      return url;
    }
  }
}
