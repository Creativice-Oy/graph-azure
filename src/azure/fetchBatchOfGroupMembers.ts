import {
  IntegrationCacheEntry,
  IntegrationStepIterationState,
} from "@jupiterone/jupiter-managed-integration-sdk";
import { Group } from "@microsoft/microsoft-graph-types";

import { AzureExecutionContext, GroupsCacheState } from "../types";
import { getBatchPages, getPageLimit } from "./batch";
import { GroupMember } from "./types";
import AzureClientError from "./AzureClientError";

export default async function fetchBatchOfGroupMembers(
  executionContext: AzureExecutionContext,
  iterationState: IntegrationStepIterationState,
): Promise<IntegrationStepIterationState> {
  const { azure, logger } = executionContext;
  const cache = executionContext.clients.getCache();
  const groupsCache = cache.iterableCache<
    IntegrationCacheEntry,
    GroupsCacheState
  >("groups");

  const batchPages = getBatchPages("group_members", 10);
  const limit = getPageLimit("group_members", 200);

  let membersCount: number = iterationState.state.count || 0;
  let nextLink: string | undefined = iterationState.state.nextLink;
  let groupIndex: number = iterationState.state.groupIndex || 0;

  let pagesProcessed = 0;
  let totalGroups = 0;
  let fetchError: AzureClientError | undefined;

  logger.info(
    { totalGroups, groupIndex },
    "Starting group iteration for members fetch",
  );

  await groupsCache.forEach(
    async (e) => {
      const groupEntry = e.entry;
      const groupEntryIndex = e.entryIndex;
      const group = groupEntry.data as Group;

      totalGroups = e.totalEntries;

      // load up accumulated list of members
      const groupMembers: GroupMember[] =
        (group.members as GroupMember[]) || [];

      logger.info(
        { totalGroups, groupEntryIndex, groupIndex },
        "Starting group members fetch loop",
      );

      // fetch members for group
      do {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const response = await azure.fetchGroupMembers(group.id!, {
          nextLink,
          limit,
          // Changes here should be reflected in `GroupMember`
          select: ["id", "displayName", "jobTitle", "mail"],
        });

        if (!response.err) {
          groupMembers.push(...response.resources);
          membersCount += response.resources.length;
          nextLink = response.nextLink;
        } else {
          fetchError = response.err;
        }

        pagesProcessed++;
      } while (!fetchError && pagesProcessed < batchPages && nextLink);

      logger.info(
        { fetchError, nextLink, pagesProcessed },
        "Completed group members fetch loop",
      );

      if (fetchError) {
        // Stop iterating groups altogether until we learn more about the errors.
        // Are there some groups we don't have access to, so that we should skip
        // them, or is this a temporary problem that should be ignored (do not
        // produce any operations to avoid unintended data deletion)?
        return true;
      } else {
        // update the cached group to include members loaded during this invocation
        await groupsCache.putEntries([
          {
            key: groupEntry.key,
            data: { ...groupEntry.data, members: groupMembers },
          },
        ]);

        // move to the next group if there are no more members in this group
        if (!nextLink) {
          groupIndex = groupEntryIndex + 1;
        }

        // stop iteration when pagesProcessed reaches invocation limit
        return pagesProcessed >= batchPages;
      }
    },
    { skip: groupIndex },
  );

  const groupMembersFetchCompleted =
    typeof nextLink !== "string" && groupIndex >= totalGroups;

  const finished = !!fetchError || groupMembersFetchCompleted;

  logger.info(
    {
      fetchError,
      nextLink,
      pagesProcessed,
      totalGroups,
      groupIndex,
      groupMembersFetchCompleted,
      finished,
    },
    "Completed group iteration for members fetch",
  );

  await groupsCache.putState({
    groupMembersFetchCompleted: !fetchError && groupMembersFetchCompleted,
    fetchError: fetchError,
  });

  return {
    ...iterationState,
    finished,
    state: {
      nextLink,
      limit,
      pages: pagesProcessed,
      count: membersCount,
      groupIndex,
    },
  };
}
