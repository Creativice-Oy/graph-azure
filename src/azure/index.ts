import createAzureWebLinker from "./createAzureWebLinker";

export { createAzureWebLinker };

export * from "./types";
