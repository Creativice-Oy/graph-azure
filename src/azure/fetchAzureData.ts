import AzureClient from "./AzureClient";
import { AzureDataModel, Group, GroupMembers } from "./types";

export default async function fetchAzureData(
  client: AzureClient,
): Promise<AzureDataModel> {
  const [groups, users] = await Promise.all([
    client.fetchGroups(),
    client.fetchUsers(),
  ]);

  const groupsMembers = await Promise.all(
    groups.map((group: Group) => fetchGroupMembers(group, client)),
  );

  return { groups, groupsMembers, users };
}

async function fetchGroupMembers(
  group: Group,
  client: AzureClient,
): Promise<GroupMembers> {
  const members = await client.fetchMembers(group.id);

  return {
    group,
    members,
  };
}
