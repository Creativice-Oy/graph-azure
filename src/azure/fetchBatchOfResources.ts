import {
  IntegrationCacheEntry,
  IntegrationStepExecutionResult,
  IntegrationStepIterationState,
} from "@jupiterone/jupiter-managed-integration-sdk";

import { AzureExecutionContext, ResourceCacheState } from "../types";
import AzureClient, {
  FetchResourcesResponse,
  PaginationOptions,
} from "./AzureClient";
import { getBatchPages, getPageLimit } from "./batch";

/**
 * An iterating execution handler that loads resources and associated data in
 * batches of pages of resources, storing the raw response data in the
 * `IntegrationCache` for later processing in another step.
 */
export default async function fetchBatchOfResources<
  T extends microsoftgraph.Entity
>(
  executionContext: AzureExecutionContext,
  iterationState: IntegrationStepIterationState,
  resourceName: string,
  fetch: (
    azure: AzureClient,
    paginationOptions: PaginationOptions,
  ) => Promise<FetchResourcesResponse<T> | undefined>,
): Promise<IntegrationStepExecutionResult> {
  const { azure } = executionContext;
  const cache = executionContext.clients.getCache();
  const resourceCache = cache.iterableCache<
    IntegrationCacheEntry,
    ResourceCacheState
  >(resourceName);

  const batchPages = getBatchPages(resourceName, 2);
  const limit = getPageLimit(resourceName, 200);

  let pagesProcessed = 0;
  let entryCount: number = iterationState.state.count || 0;
  let nextLink: string | undefined = iterationState.state.nextLink;

  do {
    const response = await fetch(azure, {
      nextLink,
      limit,
    });

    if (response) {
      entryCount = await resourceCache.putEntries(
        response.resources.map(e => ({
          key: e.id!,
          data: e,
        })),
      );
      nextLink = response.nextLink;
    }

    pagesProcessed++;
  } while (pagesProcessed < batchPages && nextLink);

  const finished = typeof nextLink !== "string";
  await resourceCache.putState({ resourceFetchCompleted: finished });

  return {
    iterationState: {
      ...iterationState,
      finished,
      state: {
        nextLink,
        limit,
        pages: pagesProcessed,
        count: entryCount,
      },
    },
  };
}
