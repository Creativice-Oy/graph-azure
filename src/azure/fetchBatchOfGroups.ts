import {
  IntegrationStepExecutionResult,
  IntegrationStepIterationState,
} from "@jupiterone/jupiter-managed-integration-sdk";

import { AzureExecutionContext } from "../types";
import fetchBatchOfResources from "./fetchBatchOfResources";

export default async function fetchBatchOfGroups(
  executionContext: AzureExecutionContext,
  iterationState: IntegrationStepIterationState,
): Promise<IntegrationStepExecutionResult> {
  return fetchBatchOfResources(
    executionContext,
    iterationState,
    "groups",
    (azure, options) => azure.fetchGroups(options),
  );
}
