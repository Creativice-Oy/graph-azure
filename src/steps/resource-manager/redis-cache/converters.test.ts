import { createAzureWebLinker } from '../../../azure';
import {
  createRedisCacheEntity,
  createRedisFirewallRuleEntity,
} from './converters';
import {
  RedisResource,
  RedisFirewallRule,
} from '@azure/arm-rediscache/esm/models';
const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createRedisCacheEntity', () => {
  test('properties transferred', () => {
    const data: RedisResource = {
      accessKeys: undefined,
      enableNonSslPort: false,
      hostName: 'keionned-j1dev-redis-cache.redis.cache.windows.net',
      id:
        '/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache',
      instances: [
        { isMaster: true, sslPort: 15000 },
        { isMaster: false, sslPort: 15001 },
      ],
      linkedServers: [],
      location: 'East US',
      minimumTlsVersion: '1.2',
      name: 'keionned-j1dev-redis-cache',
      port: 6379,
      provisioningState: 'Succeeded',
      redisConfiguration: {
        maxclients: '2000',
        'maxfragmentationmemory-reserved': '125',
        'maxmemory-delta': '100',
        'maxmemory-reserved': '100',
      },
      redisVersion: '4.0.14',
      sku: {
        capacity: 2,
        family: 'C',
        name: 'Standard',
      },
      sslPort: 6380,
      tags: {},
      type: 'Microsoft.Cache/Redis',
    };

    const redisCacheEntity = createRedisCacheEntity(webLinker, data);

    expect(redisCacheEntity).toMatchSnapshot();
    expect(redisCacheEntity).toMatchGraphObjectSchema({
      _class: ['Database', 'DataStore', 'Cluster'],
      schema: {},
    });
  });
});

describe('createRedisFirewallRuleEntity', () => {
  test('properties transferred', () => {
    const data: RedisFirewallRule = {
      id:
        '/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache/firewallRules/j1dev_redis_cache_firewall_rule',
      name: 'keionned-j1dev-redis-cache/j1dev_redis_cache_firewall_rule',
      startIP: '1.2.3.4',
      endIP: '2.3.4.5',
      type: 'Microsoft.Cache/Redis/firewallRules',
    };

    const redisFirewallRuleEntity = createRedisFirewallRuleEntity(
      webLinker,
      data,
    );

    expect(redisFirewallRuleEntity).toMatchSnapshot();
    expect(redisFirewallRuleEntity).toMatchGraphObjectSchema({
      _class: ['Rule'],
      schema: {},
    });
  });
});
