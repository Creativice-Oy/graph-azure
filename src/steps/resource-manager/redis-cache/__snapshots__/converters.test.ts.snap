// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRedisCacheEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Database",
    "DataStore",
    "Cluster",
  ],
  "_key": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "accessKeys": undefined,
        "enableNonSslPort": false,
        "hostName": "keionned-j1dev-redis-cache.redis.cache.windows.net",
        "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache",
        "instances": Array [
          Object {
            "isMaster": true,
            "sslPort": 15000,
          },
          Object {
            "isMaster": false,
            "sslPort": 15001,
          },
        ],
        "linkedServers": Array [],
        "location": "East US",
        "minimumTlsVersion": "1.2",
        "name": "keionned-j1dev-redis-cache",
        "port": 6379,
        "provisioningState": "Succeeded",
        "redisConfiguration": Object {
          "maxclients": "2000",
          "maxfragmentationmemory-reserved": "125",
          "maxmemory-delta": "100",
          "maxmemory-reserved": "100",
        },
        "redisVersion": "4.0.14",
        "sku": Object {
          "capacity": 2,
          "family": "C",
          "name": "Standard",
        },
        "sslPort": 6380,
        "tags": Object {},
        "type": "Microsoft.Cache/Redis",
      },
    },
  ],
  "_type": "azure_redis_cache",
  "classification": null,
  "createdOn": undefined,
  "displayName": "keionned-j1dev-redis-cache",
  "enableNonSslPort": false,
  "encrypted": null,
  "hostName": "keionned-j1dev-redis-cache.redis.cache.windows.net",
  "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache",
  "location": "East US",
  "minimumTlsVersion": "1.2",
  "name": "keionned-j1dev-redis-cache",
  "port": 6379,
  "provisioningState": "Succeeded",
  "redisVersion": "4.0.14",
  "sslPort": 6380,
  "type": "Microsoft.Cache/Redis",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache",
}
`;

exports[`createRedisFirewallRuleEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Rule",
  ],
  "_key": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache/firewallRules/j1dev_redis_cache_firewall_rule",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "endIP": "2.3.4.5",
        "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache/firewallRules/j1dev_redis_cache_firewall_rule",
        "name": "keionned-j1dev-redis-cache/j1dev_redis_cache_firewall_rule",
        "startIP": "1.2.3.4",
        "type": "Microsoft.Cache/Redis/firewallRules",
      },
    },
  ],
  "_type": "azure_firewall_rule",
  "createdOn": undefined,
  "displayName": "keionned-j1dev-redis-cache/j1dev_redis_cache_firewall_rule",
  "endIp": "2.3.4.5",
  "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache/firewallRules/j1dev_redis_cache_firewall_rule",
  "name": "keionned-j1dev-redis-cache/j1dev_redis_cache_firewall_rule",
  "startIp": "1.2.3.4",
  "type": "Microsoft.Cache/Redis/firewallRules",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.Cache/Redis/keionned-j1dev-redis-cache/firewallRules/j1dev_redis_cache_firewall_rule",
}
`;
